### PR TITLE
FIX : #13968

### PR DESCRIPTION
**Pull request by FHenry**
FIX : #13968

### CODE REVIEW

**Comment by eldy** [comment]
I don't like the implementation to solve this bug. It introduces another complex setup that we can avoid. I suggest this better change: We can and we should rely only on the array childtablesoncascade
Instead of providing the name of a table into this array, we can provide a name of a class prefixed with "@" to make a difference between table name and class name.
So childtablesoncascade become

childtablesoncascade = array ('table1', 'table2', '@objectname:classpath') 

instead of 

childtablesoncascade = array ('table1', 'table2', 'table3')

The into the look tthat use the childtablesoncascade, if table name start with @ , we init the class, fetch and call delete().

So the second level of table to delete will be done by the childtablesoncascade of first level. And declaration of child will always be limited to direct child only. This is also better as this can execute triggers of sublevel and is able to manage infinite number of level and we maintain the child list only in one place (no need to maintain each grand parent to each time we add a table).

**Comment by FHenry** [comment]
@eldy You'r right, I prefer your version. 
I've change my commit to code as you said. The only thing that hit me is variable name "childtabletodelete" that do not contains only table name, but it's explain in comment.
At least it's clean like that and good Triggers (and all others childs object contrainst deletion) are executed when a thirdparty and contact are deleteed (taht wasn't the case before).
I was thinking about transfrom all child table into object, but may be it's overkill.

File: htdocs/contact/class/contact.class.php [hunk 1]
```diff
@@ -1731,4 +1731,57 @@ public function updateRoles()
...
 	}
+
+	/**
+	 * Delete all contact from a thirdparty
+	 * @param		int		$socId      Thirdparty Id
+	 * @param       int     $notrigger  Disable all trigger
+	 * @return		int						<0 if KO, >0 if OK
+	 * @throws Exception
+	 */
+	public function deleteBySoc($socId = 0, $notrigger = 0)
```
Comment (eldy) [inline htdocs/contact/class/contact.class.php:1742]: I thought to something even better. We can still reduce code:

1) Rename deleteBySoc into deleteByParentField (so name work for child object) and add a parameter $parentfield = 'fk_soc')

2) Move the method into commonobject.class.php (so method is available for every class)

3) Into definition of childtablesoncascade for societe, you can replace deleteBySoc() method (Name will always be deleteByParentField so we don't need it) with 'fk_soc'. Provide value of this field name as a new param when calling 
$childObject->{$deleteMethod}($this->id) => It become  
$childObject->deleteByParentField($this->id, 'fk_soc')

So A) there is no more dynamic method call (we should avoid if we can) and B) solution will works for all child object (this was my first thinking) without adding code and C) deleteByParentField can be used for a lot of other use cases.

Comment (FHenry) [reply htdocs/contact/class/contact.class.php:1742]: Ok, i'm on it

**Comment by FHenry** [comment]
@eldy what do you think about this last code review ?

File: htdocs/contact/class/contact.class.php (hunk 1 above)
Comment (FHenry) [reply htdocs/contact/class/contact.class.php]: new commit pushed

**Comment by FHenry** [comment]
@eldy, so in php 7.4 it will not work at all, we send $user (Unser class instance) onto an int for contact->delete($notrigger)....

Why remove the reflexion test ?

**Comment by eldy** [comment]
Using "Reflection" for production running code is not a good practice.
- If you need to make some tests because some code is comptabile with vx and not v8, it is better to have a test if ((float) DOL_VERSION <= x) 
- If you need to make different call of a method because code is not correctly standardized, it is better to have a test
if (get_class($this) == 'Contact) {  // Special code because delete for contact has not been standardized like other delete.

So impact analysis is better, tracking on version regression is also better and you also don't break code quality tools due to "code that generate or execute code depending on code". I added the test for v11 but good change will be to standardize delete() in develop where first param should be $user everywhere.

**Pull request merged**